### PR TITLE
New gallery examples

### PR DESCRIPTION
**Pull request by ehsteve**
A bunch of new gallery examples

### CODE REVIEW

**Comment by wafels** [comment]
No fitting of histograms with line fitting programs!  Nope.  Wrong way to do this. If you want to fit a distribution to your data do a maximum likelihood fit of all the data - no binning.

File: examples/mask_disk.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+###############################################################################
+# Next we build two arrays which include all of the x and y pixel indices.
+# We must not forget to add the correct units because we will next pass this_sji_map
+# into a sunpy function which requires them.
+x, y = np.meshgrid(*[np.arange(v.value) for v in aia.dimensions]) * u.pixel
+
+###############################################################################
+# Now we can convert this to helioprojective coordinates and create a new
+# array which contains the normalized radial position for each pixel
+hpc_x, hpc_y = aia.pixel_to_data(x,y)
```
Comment (larrymanley) [inline examples/mask_disk.py]: gives 

> ValueError: too many values to unpack (expected 2)

File: examples/mask_disk.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from sunpy.map import Map
+###############################################################################
+# We first create the Map using the sample data.
+aia = Map(AIA_171_IMAGE)
+
+###############################################################################
+# Next we build two arrays which include all of the x and y pixel indices.
+# We must not forget to add the correct units because we will next pass this_sji_map
+# into a sunpy function which requires them.
+x, y = np.meshgrid(*[np.arange(v.value) for v in aia.dimensions]) * u.pixel
```
Comment (larrymanley) [inline examples/mask_disk.py]: astropy.units import missing at top

**Comment by ehsteve** [comment]
`off_limb_enhance.py`
gallery_offlimb_dependence
gallery_offlimb_enhance

**Comment by ehsteve** [comment]
`mask_disk.py`
mask_disk

**Comment by ehsteve** [comment]
`loop_edge_enhance.py`
gallery_sharp

File: examples/loop_edge_enhance.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+sx = ndimage.sobel(aia_smap.data, axis=0, mode='constant')
+sy = ndimage.sobel(aia_smap.data, axis=1, mode='constant')
+edge_enhanced_im = np.hypot(sx, sy)
+
+###############################################################################
+# Finally we create a new map with the edge enhanced data and plot the result.
+edge_map = Map(edge_enhanced_im, aia_smap.meta)
+
+plt.figure(figsize=(12, 4))
+plt.subplot(121)
```
Comment (Cadair) [inline examples/loop_edge_enhance.py]: `projection=aia_smap`

File: examples/loop_edge_enhance.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+aia_smap.plot()
+plt.subplot(122)
```
Comment (Cadair) [inline examples/loop_edge_enhance.py]: `projection=edge_map`

File: examples/loop_edge_enhance.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+edge_map.plot()
+plt.savefig("gallery_sharp.png")
+#plt.show()
```
Comment (Cadair) [inline examples/loop_edge_enhance.py]: uncomment

File: examples/map_data_histogram.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+Map Histogram
+=============
+
+How to plot and fit the histogram of the data of a map.
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+from sunpy.data.sample import AIA_171_IMAGE
+from sunpy.map import Map
+from astropy.modeling import models, fitting
```
Comment (Cadair) [inline examples/map_data_histogram.py]: import ordering and line breaks.

File: examples/map_data_histogram.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+=============
+Map Histogram
+=============
+
+How to plot and fit the histogram of the data of a map.
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+from sunpy.data.sample import AIA_171_IMAGE
+from sunpy.map import Map
```
Comment (Cadair) [inline examples/map_data_histogram.py]: import `sunpy.map`

File: examples/map_data_histogram.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+from astropy.modeling import models, fitting
+
+###############################################################################
+# We first create the Map using the sample data.
+aia = Map(AIA_171_IMAGE)
```
Comment (Cadair) [inline examples/map_data_histogram.py]: `sunpy.map.Map`

File: examples/mask_disk.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+======================
+Masking the solar disk
+======================
+
+This example shows how to mask off emission from the disk.
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+import numpy.ma as ma
+import astropy.units as u
```
Comment (Cadair) [inline examples/mask_disk.py]: import line breaks

File: examples/mask_disk.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+from sunpy.data.sample import AIA_171_IMAGE
+from sunpy.map import Map
```
Comment (Cadair) [inline examples/mask_disk.py]: `import sunpy.map`

File: examples/mask_disk.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+###############################################################################
+# Next we build two arrays which include all of the x and y pixel indices.
+# We must not forget to add the correct units because we will next pass this_sji_map
+# into a sunpy function which requires them.
+x, y = np.meshgrid(*[np.arange(v.value) for v in aia.dimensions]) * u.pixel
+
+###############################################################################
+# Now we can convert this to helioprojective coordinates and create a new
+# array which contains the normalized radial position for each pixel
+hpc_coords = aia.pixel_to_data(x,y)
```
Comment (Cadair) [inline examples/mask_disk.py]: space after comma

File: examples/off_limb_enhance.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
+"""
+===========================
+Enhancing Off-limb emission
+===========================
+
+This example shows how to enhance emission above the limb.
+"""
+import numpy as np
+import astropy.units as u
```
Comment (Cadair) [inline examples/off_limb_enhance.py]: import ordering is not wrong but is inconsistent with the other examples where you have imported astropy as a separate block.

File: examples/off_limb_enhance.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+
+from sunpy.data.sample import AIA_171_IMAGE
+from sunpy.map import Map
```
Comment (Cadair) [inline examples/off_limb_enhance.py]: `import sunpy.map`

File: examples/off_limb_enhance.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+###############################################################################
+# Next we build two arrays which include all of the x and y pixel indices.
+# We must not forget to add the correct units because we will next pass this
+# into a function which requires them.
+x, y = np.meshgrid(*[np.arange(v.value) for v in aia.dimensions]) * u.pix
+
+###############################################################################
+# Now we can convert this to helioprojective coordinates and create a new
+# array which contains the normalized radial position for each pixel
+hpc_coords = aia.pixel_to_data(x,y)
```
Comment (Cadair) [inline examples/off_limb_enhance.py]: space after comma

File: examples/off_limb_enhance.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+# Let's now plot and compare the results.
+scaled_map = Map(aia.data * scale_factor, aia.meta)
+norm = colors.Normalize(vmin=10, vmax=10000)
+plt.figure(figsize=(12,5))
+plt.subplot(1,2,1)
+aia.plot()
+plt.subplot(1,2,2)
+scaled_map.plot()
+plt.savefig('gallery_offlimb_enhance.png')
+#plt.show()
```
Comment (Cadair) [inline examples/off_limb_enhance.py]: uncomment

File: examples/plot_sample_images.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+=============
+Sample Images
+=============
+
+This example shows off all of the sample images provided by SunPy.
+"""
+import matplotlib.pyplot as plt
+
+from sunpy.data.sample import file_dict
+from sunpy.map import Map
```
Comment (Cadair) [inline examples/plot_sample_images.py]: `import sunpy.map`

File: examples/plot_sample_images.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+# We create a loop through all of the sample shortcuts and picks out the images
+# that can be opened by Map. It then plots each one.
+for key in file_dict:
+    if key.count('IMAGE'):
+        try:
+            m = Map(file_dict[key])
+            plt.figure()
+            m.plot()
+            plt.savefig('{0}.png'.format(key))
+            #plt.show()
```
Comment (Cadair) [inline examples/plot_sample_images.py]: move to end of file and uncomment.

File: examples/searching_vso.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+
+A simple example showing how to download data from the VSO with Fido.
+"""
+
+###############################################################################
+# Fido is the primary interface to search for and download data and
+# will search the VSO if appropriate. First import it and the search
+# attributes.
+import astropy.units as u
+from sunpy.net import Fido, attrs as a
```
Comment (Cadair) [inline examples/searching_vso.py]: import spacing

**Comment by Cadair** [comment]
These examples are awesome! Just what we should have the gallery full of, thanks @ehsteve

**Comment by ehsteve** [comment]
`lat_lon_lines.py`
lonlat0

File: examples/lat_lon_lines.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+# of zero latitude. First define the coordinates as we did before and then
+# transform them.
+lon0 = SkyCoord(np.linspace(-80, 80, 100) * u.deg,
+                np.zeros(100) * u.deg, frame=frames.HeliographicStonyhurst)
+lat0 = SkyCoord(np.zeros(100)*u.deg,
+                np.linspace(-90, 90, 100)*u.deg, frame=frames.HeliographicStonyhurst)
+
+hpc_lon0 = lon0.transform_to(aia.coordinate_frame)
+hpc_lat0 = lat0.transform_to(aia.coordinate_frame)
+
```
Comment (Cadair) [inline examples/lat_lon_lines.py]: To make this about slightly more that just drawing (which could be done with `draw_grid` or `WCSAxes`) you could add a section extracting the image values along these lines.

**Comment by ehsteve** [comment]
Updated the histogram example to create the following two plots.
figure_hist
figure_1

**Comment by ehsteve** [comment]
ping @wafels

**Comment by ehsteve** [comment]
FYI, @Cadair, I won't be add have time to add more examples now so I'll work on addressing comments.

**Comment by ehsteve** [comment]
I therefore removed the WIP tag.

File: examples/lat_lon_lines.py [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+import numpy as np
+import astropy.units as u
+import matplotlib.pyplot as plt
+from sunpy.data.sample import AIA_171_IMAGE
+import sunpy.map
+
+###############################################################################
+# We first create the Map using the sample data and import the coordinate
+# functionality.
+from astropy.coordinates import SkyCoord
```
Comment (Cadair) [inline examples/lat_lon_lines.py]: imports should be in the top block?

File: examples/mask_disk.py [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+This example shows how to mask off emission from the disk.
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+import numpy.ma as ma
+import astropy.units as u
+from sunpy.data.sample import AIA_171_IMAGE
+import sunpy.map
+###############################################################################
```
Comment (Cadair) [inline examples/mask_disk.py]: line break above here.

File: examples/off_limb_enhance.py [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+plt.plot(rsun_array, y, label='data')
+label = r'fit=Ax^{:.2f}'.format(params[0])
+plt.plot(rsun_array, 10**np.poly1d(params)(rsun_array), label=label)
+plt.yscale('log')
+plt.ylabel('mean DN')
+plt.xlabel('Rsun')
+plt.legend()
+plt.show()
+###############################################################################
```
Comment (Cadair) [inline examples/off_limb_enhance.py]: line break above.

File: examples/off_limb_enhance.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+===========================
+
+This example shows how to enhance emission above the limb.
+"""
+import numpy as np
+import astropy.units as u
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+
+from sunpy.data.sample import AIA_171_IMAGE
```
Comment (Cadair) [inline examples/off_limb_enhance.py]: this line break is inconsistent with the other import blocks, personally I like line breaks in imports, but consistency is important.

File: examples/loop_edge_enhance.py [hunk 24]
```diff
@@ -0,0 +1,39 @@
+"""
+==============================
+Edge Enhancing Filtering a Map
+==============================
+
+This example shows how to edge enhance a coronal loops in an SDO/AIA image.
```
Comment (wafels) [inline examples/loop_edge_enhance.py]: remove "a" from "enhance a coronal"

**Comment by wafels** [comment]
In  examples/lat_lon_lines.py, it might be worth mentioning that GreatArc can be used to draw lines on the sphere of the Sun between any two arbitrary points.

**Comment by Cadair** [comment]
@ehsteve doc build fail is real: https://travis-ci.org/sunpy/sunpy/jobs/260277269#L1857

File: examples/lat_lon_lines.py (hunk 20 above)
Comment (nabobalis) [inline examples/lat_lon_lines.py]: Well isn't this more of a walkthrough, so there is no problem with these imports here since its linked to the text.

**Comment by nabobalis** [comment]
The doc build was real, I "fixed" by copy and pasting code I dont understand. 
The example was ```masking_hot_pixels.py```.

File: examples/lat_lon_lines.py (hunk 20 above)
Comment (Cadair) [inline examples/lat_lon_lines.py]: well in my opinion if they are going to be here this comment should be about the imports. Talking about how sunpy coordinates uses astropy etc. or they should be in the top block.

Comment (nabobalis) [inline examples/lat_lon_lines.py]: FINE

File: examples/lat_lon_lines.py [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+hpc_lon0 = lon0.transform_to(aia.coordinate_frame)
+hpc_lat0 = lat0.transform_to(aia.coordinate_frame)
+
+###############################################################################
+# Let's now plot the results. We'll overlay the autogenerated SunPy lon/lat
+# grid as well for comparison.
+fig = plt.figure()
+ax = plt.subplot(projection=aia)
+aia.plot()
+ax.plot(hpc_lon0.Tx.to('deg'), hpc_lon0.Ty.to('deg'),
```
Comment (Cadair) [inline examples/lat_lon_lines.py]: this should use `ax.plot_coord`

File: examples/lat_lon_lines.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+        transform=ax.get_transform('world'), color='blue')
+ax.plot(hpc_lat0.Tx.to('deg'), hpc_lat0.Ty.to('deg'),
```
Comment (Cadair) [inline examples/lat_lon_lines.py]: as should this

File: examples/masking_hot_pixels.py [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+mask = ma.masked_greater_equal(r, 1)
+scaled_map = smap.Map(aia.data, aia.meta, mask=mask.mask)
+
+###############################################################################
+# Let's now plot the results. We'll overlay the autogenerated SunPy lon/lat
+# grid as well for comparison.
+fig = plt.figure()
+ax = plt.subplot(projection=aia)
+aia.plot()
+ax.plot(hpc_max.Tx.to('deg'), hpc_max.Ty.to('deg'), 'bx',
```
Comment (Cadair) [inline examples/masking_hot_pixels.py]: should be `ax.plot_coord`

**Comment by Cadair** [comment]
This has been rebased.

File: examples/lat_lon_lines.py [hunk 28]
```diff
@@ -48,9 +48,7 @@
 fig = plt.figure()
 ax = plt.subplot(projection=aia)
 aia.plot()
-ax.plot(hpc_lon0.Tx.to('deg'), hpc_lon0.Ty.to('deg'),
-        transform=ax.get_transform('world'), color='blue')
-ax.plot(hpc_lat0.Tx.to('deg'), hpc_lat0.Ty.to('deg'),
-        transform=ax.get_transform('world'), color='blue')
+ax.plot_coord(hpc_lat0, color="C0")
+ax.plot_coord(hpc_lon0, color="C0")
```
Comment (Cadair) [inline examples/lat_lon_lines.py]: @ehsteve please appreciate the beauty of this ;)

File: examples/brightness_pixel_location.py [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+
+hpc_max = aia.pixel_to_data(max_indices[1], max_indices[0])
+
+###############################################################################
+# Let's now plot the results. We'll overlay the autogenerated SunPy lon/lat
+# grid as well for comparison.
+fig = plt.figure()
+ax = plt.subplot(projection=aia)
+aia.plot()
+ax.plot(hpc_max.Tx.to('deg'), hpc_max.Ty.to('deg'), 'bx',
```
Comment (nabobalis) [inline examples/brightness_pixel_location.py]: Does this need plot_coord? @Cadair

Comment (Cadair) [inline examples/brightness_pixel_location.py]: Yes

**Pull request merged**